importer/helpers/helpers.go

```diff
@@ -113,8 +113,9 @@ func (n *UnixfsNode) AddChild(child *UnixfsNode, db *DagBuilderHelper) error {
 	}
 
 	// Pin the child node indirectly
-	if db.mp != nil {
-		db.mp.PinWithMode(childkey, pin.Indirect)
+	err = db.bcb(childkey, false)
```

Comment:
On Wed, May 27, 2015 at 12:13:21PM -0700, Juan Batiz-Benet wrote:
> >  Couldn't we just handle root-level changes in the caller?
> 
> Yes, but @whyrusleeping's point with this PR is to not do things at
> the caller / per-root because it's very expensive (and unsafe) to
> pin everything after the fact.  (or do you mean something
> different?)

I mean that [1] could happen in the Add()-caller, instead of having a
single true-root callback at the end of Add() [2].  That's the same
computational cost (just shifting where root-specific code is stored).
It's maybe more work, if lots of Add()-callers want the same root-node
handling, but I'd argue that the way to handle that would be to make
that common handling easy to apply, instead of overloading the
post-new-DAG-node callback.

[1]: https://github.com/ipfs/go-ipfs/commit/8dae027aebee07079c137f1b80e758318e2513d1#diff-1de672191acc5920abf8bffeec0beb73R70
[2]: https://github.com/ipfs/go-ipfs/commit/8dae027aebee07079c137f1b80e758318e2513d1#diff-fdae772d2b35bc89d019b608d02c1755R145